tests/sentry/workflow_engine/models/test_detector.py

```diff
@@ -0,0 +1,119 @@
+import unittest
+
+from sentry.testutils.cases import TestCase
+from sentry.workflow_engine.models import DataPacket, DetectorEvaluationResult
+from sentry.workflow_engine.models.detector import (
+    Detector,
+    DetectorStateData,
+    StatefulDetectorHandler,
+    get_redis_client,
+)
+from sentry.workflow_engine.models.detector_state import DetectorState, DetectorStatus
+
+
+class MockDetectorStateHandler(StatefulDetectorHandler[dict]):
+    def evaluate(self, data_packet: DataPacket[dict]) -> list[DetectorEvaluationResult]:
+        return []
+
+
+class TestKeyBuilders(unittest.TestCase):
+    def build_handler(self, detector: Detector | None = None) -> MockDetectorStateHandler:
+        if detector is None:
+            detector = Detector(id=123)
+        return MockDetectorStateHandler(detector)
+
+    def test(self):
+        assert self.build_handler().build_dedupe_value_key("test") == "123:test:dedupe_value"
+        assert self.build_handler().build_counter_value_key("test", "name_1") == "123:test:name_1"
+
+    def test_different_dedupe_keys(self):
+        handler = self.build_handler()
+        handler_2 = self.build_handler(Detector(id=456))
+        assert handler.build_dedupe_value_key("test") != handler_2.build_dedupe_value_key("test")
+        assert handler.build_dedupe_value_key("test") != handler_2.build_dedupe_value_key("test2")
+        assert handler.build_dedupe_value_key("test") == handler.build_dedupe_value_key("test")
+        assert handler.build_dedupe_value_key("test") != handler.build_dedupe_value_key("test_2")
+
+    def test_different_counter_value_keys(self):
+        handler = self.build_handler()
+        handler_2 = self.build_handler(Detector(id=456))
+        assert handler.build_counter_value_key(
+            "test", "name_1"
+        ) != handler_2.build_counter_value_key("test", "name_1")
+        assert handler.build_counter_value_key("test", "name_1") == handler.build_counter_value_key(
+            "test", "name_1"
+        )
+        assert handler.build_counter_value_key("test", "name_1") != handler.build_counter_value_key(
+            "test2", "name_1"
+        )
+        assert handler.build_counter_value_key("test", "name_1") != handler.build_counter_value_key(
+            "test", "name_2"
+        )
+        assert handler.build_counter_value_key("test", "name_1") != handler.build_counter_value_key(
+            "test2", "name_2"
+        )
+
+
+class TestCommitStateUpdateData(TestCase):
+    def setUp(self):
+        super().setUp()
+
+    def build_handler(self, detector: Detector | None = None) -> MockDetectorStateHandler:
+        if detector is None:
+            detector = self.create_detector()
+        return MockDetectorStateHandler(detector)
+
+    def test(self):
```

Comment:
are there any common failure scenarios that we should ensure works as well?

Reply:
The main scenario is that there's no registered handler. Most other things would be database connection related. There's maybe a discussion to be had on what to do on failures - but generally, in consumers we choose to drop on failure to make sure that the consumer isn't blocked